### PR TITLE
[DHCP] Support dynamic IP addresses for the app tier

### PR DESCRIPTION
**Pull request by KimForss**
## Problem
Currently the solution requires fixed IP addresses for application VMs

## Solution
Add a property in the input json that controls the IP adressing with the default value of false

dynamic_ipaddresses = try(var.application.dynamic_addressing, false)

private_ip_address_allocation = local.dynamic_ipaddresses ? "Dynamic" : "Static"

## Tests
Add "dynamic_addressing" : true, in the application section in the json

## Notes

This change will "break" the pipeline as the ip addresses in feature-tfstate2 are static and this deployment will change the IPs

### CODE REVIEW

File: deploy/terraform/run/sap_system/output.tf [hunk 1]
```diff
@@ -0,0 +1,7 @@
+output "dns_information_app" {
+    value = module.app_tier.dns_info_vms
+}
+
+output "dns_information_loadbalancers_app" {
+    value = module.app_tier.dns_info_loadbalancers
+}
```
Comment (rukawata) [inline deploy/terraform/run/sap_system/output.tf:7]: nit: newline :)

File: deploy/terraform/run/sap_system/sapsystem_full.json [hunk 2]
```diff
@@ -63,6 +63,7 @@
     "app_zones": ["1","2"],
     "scs_zones": ["1"],
     "web_zones": ["1"],
+    "dynamic_addressing" : false,
```
Comment (rukawata) [inline deploy/terraform/run/sap_system/sapsystem_full.json:66]: Can we call it `use_DHCP` or `enable_DHCP` instead of `dynamic_addressing`?

File: deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf [hunk 3]
```diff
@@ -137,7 +137,7 @@ locals {
 
   //Anchor VM
   anchor                      = try(local.var_infra.anchor_vms, {})
-  deploy_anchor               = length(local.anchor) > 0 ? true : false
+  deploy_anchor               = length(local.anchor)  > 0 && local.enable_db_deployment ? true : false
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf:140]: This seems to be a bug that is not related to DHCP. Can we have a small PR to just fix that?

File: deploy/terraform/terraform-units/modules/sap_system/app_tier/variables_local.tf [hunk 4]
```diff
@@ -167,6 +167,8 @@ locals {
   web_nic_ips              = try(var.application.web_nic_ips, [])
   web_admin_nic_ips        = try(var.application.web_admin_nic_ips, [])
 
+  dynamic_ipaddresses = try(var.application.dynamic_addressing, false)
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/app_tier/variables_local.tf:170]: Can we change the var name accordingly as per other comment? eg. `enable_DHCP`.

**Review by rukawata — COMMENTED** [review]
Overall no big issue. Please check the comments. Thanks!

File: deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf (hunk 3 above)
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf:140]: https://github.com/Azure/sap-hana/pull/951

File: deploy/terraform/run/sap_system/output.tf (hunk 1 above)
Comment (KimForss) [reply deploy/terraform/run/sap_system/output.tf:7]: done

File: deploy/terraform/run/sap_system/sapsystem_full.json (hunk 2 above)
Comment (KimForss) [reply deploy/terraform/run/sap_system/sapsystem_full.json:66]: "use_DHCP" is better

File: deploy/terraform/terraform-units/modules/sap_system/app_tier/variables_local.tf (hunk 4 above)
Comment (KimForss) [reply deploy/terraform/terraform-units/modules/sap_system/app_tier/variables_local.tf:170]: done

File: deploy/terraform/run/sap_system/output.tf (hunk 1 above)
Comment (rukawata) [reply deploy/terraform/run/sap_system/output.tf:7]: I still see it... can you double check?

File: deploy/terraform/terraform-units/modules/sap_system/app_tier/outputs.tf [hunk 5]
```diff
@@ -61,3 +61,71 @@ output "ers_lb_ip" {
 output "application" {
   value = local.application
 }
+
+// Output for DNS
+output "dns_info_vms" {
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/app_tier/outputs.tf:66]: Is it possible to make it just a list of vm: ip key pairs, rather than a dictionary?

eg:
```
dns_information_app = [
    "U5847-EAUS-SAP-PRD_prdapp00l4b5" = "10.11.57.110",
    "U5847-EAUS-SAP-PRD_prdscs00l4b5" = "10.11.57.106",
    "U5847-EAUS-SAP-PRD_prdweb00l4b5" = "10.11.57.125"
]
dns_information_loadbalancers_app = [
  {
    "U5847-EAUS-SAP-PRD_scs" = "10.11.57.101"
  },
  {
    "U5847-EAUS-SAP-PRD_ers" = "10.11.57.102"
  },
  {
    "U5847-EAUS-SAP-PRD_web-alb" = "10.11.57.126"
  },
]

```

**Comment by KimForss** [comment]
Yes, that would be better

________________________________

File: deploy/terraform/terraform-units/modules/sap_system/app_tier/variables_local.tf [hunk 6]
```diff
@@ -404,4 +406,17 @@ locals {
...
+    format("%s%s%s%s", local.prefix, local.separator, vm, local.resource_suffixes.vm)]
+  )
+
+  full_scsserver_names = flatten([for vm in local.scs_virtualmachine_names :
+    format("%s%s%s%s", local.prefix, local.separator, vm, local.resource_suffixes.vm)]
+  )
+
+  full_webserver_names = flatten([for vm in local.web_virtualmachine_names :
+    format("%s%s%s%s", local.prefix, local.separator, vm, local.resource_suffixes.vm)]
+  )
```
Comment (rukawata) [inline deploy/terraform/terraform-units/modules/sap_system/app_tier/variables_local.tf:421]: These changes breaks the deployment. Can you check?

**Comment by KimForss** [comment]
yes – my bad – just noticed it

**Review by rukawata — APPROVED** [review]
Thanks for addressing he comments!
